[FLINK-3544] Introduce ResourceManager component

mxm opened the pull request:
So far the JobManager has been the central instance which is responsible for resource management and allocation.

While thinking about how to integrate Mesos support in Flink, people from the Flink community realized that it would be nice to delegate resource allocation to a dedicated process. This process may run independently of the JobManager which is a requirement for proper integration of cluster allocation frameworks like Mesos.

This has led to the idea of creating a new component called the ResourceManager. Its task is to allocate and maintain resources requested by the JobManager. The ResourceManager has a very abstract notion of resources.

Initially, we thought we could make the ResourceManager deal with resource allocation and the registration/supervision of the TaskManagers. However, this approach proved to add unnecessary complexity to the runtime. Registration state of TaskManagers had to be kept in sync at both the JobManager and the ResourceManager.

That's why @StephanEwen and me changed the ResourceManager's role to simply deal with the resource acquisition. The TaskManagers still register with the JobManager which informs the ResourceManager about the successful registration of a TaskManager. The ResourceManager may inform the JobManager of failed TaskManagers. Due to the insight which the ResourceManager has in the resource health, it may detect failed TaskManagers much earlier than the heartbeat-based monitoring of the JobManager.

At this stage, the ResourceManager is an optional component. That means the JobManager doesn't depend on the ResourceManager as long as it has enough resources to perform the computation. All bookkeeping is performed by the JobManager. When the ResourceManager connects to the JobManager, it receives the current resources, i.e. task manager instances, and allocates more containers if necessary. The JobManager adjusts the number of containers through the SetWorkerPoolSize method.

In standalone mode, the ResourceManager may be deactivated or simply use the StandaloneResourceManager which does practically nothing because we don't need to allocate resources in standalone mode.

In YARN mode, the ResourceManager takes care of communicating with the Yarn resource manager. When containers fail, it informs the JobManager and tries to allocate new containers. The ResourceManager runs as an actor within the same actor system as the JobManager. It could, however, also run independently. The independent mode would be the default behavior for Mesos where the framework master is expected to just deal with resource allocation.

The attached figures depict the message flow between ResourceManager, JobManager, and TaskManager.

[ResourceManagerSketch.pdf](https://github.com/apache/flink/files/151425/ResourceManagerSketch.pdf)

mxm commented:
Our testing infrastructure is currently very unstable. Here is at least a build where everything passes: https://travis-ci.org/mxm/flink/builds/115046589

If nobody objects, I would like to merge the changes tomorrow.

rmetzger commented:
I'll review the PR now

rmetzger commented on flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java:
```diff
@@ -240,6 +246,37 @@
+	 * Similar to the {@see CONTAINERED_MASTER_ENV_PREFIX}, this configuration prefix allows
+	 * setting custom environment variables for the workers (TaskManagers)
+	 */
+	public static final String CONTAINERED_TASK_MANAGER_ENV_PREFIX = "containered.taskmanager.env.";
```
I wonder if we should prefix these config params with a `jobmanager.resource-framework` prefix?

mxm replied:
I'll revise these names.

mxm replied:
I thought about theses prefixes again. I think they make sense but we could possibly change them before the release in a follow-up.

rmetzger commented on flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java:
```diff
@@ -682,27 +729,27 @@
+	 * heap size
 	 */
-	public static final float DEFAULT_YARN_HEAP_CUTOFF_RATIO = 0.25f;
+	public static final float DEFAULT_CONTAINERED_HEAP_CUTOFF_RATIO = 0.25f;
```
You can not rename variables in the ConfigConstants because the class is publicly visible.

mxm replied:
I see. Will revert the change.

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/BootstrapTools.java:
```diff
@@ -0,0 +1,344 @@
+			}
+
+			ActorSystem actorSystem = AkkaUtils.createActorSystem(akkaConfig);
+			logger.info("Actor system started at " + hostPortUrl);
```
we usually use info("some log string {}", value); to avoid string concat when logging is disabled.

mxm replied:
That's how I prefer it as well. Must have been copied over when the helper functions were created.

rmetzger replied:
No problem ;)

rmetzger commented on flink-runtime/src/main/scala/org/apache/flink/runtime/minicluster/FlinkMiniCluster.scala:
```diff
@@ -366,6 +428,28 @@ abstract class FlinkMiniCluster(
+//
+//    Await.ready()
+//
+//  }
```
Why is this commented out?

mxm replied:
Good catch. Commented this utility function because it was not necessary anymore.

rmetzger commented:
I tried finding updates to the documentation, but I couldn't find any (maybe its because of the PR size). We should at least add the new configuration parameters.

rmetzger commented:
I tested the change on a secured CDH 5.3 setup and it worked, also across TaskManager failures:

```
2016-03-11 05:27:34,732 INFO  org.apache.flink.yarn.YarnJobManager                          - Status of job 58bd9ea7c4c4515ecd1cb29928523662 (Flink Java Job at Fri Mar 11 05:27:33 PST 2016) changed to FINISHED.
2016-03-11 05:28:16,832 WARN  akka.remote.ReliableDeliverySupervisor                        - Association with remote system [akka.tcp://flink@127.0.0.1:48995] has failed, address is now gated for [5000] ms. Reason is: [Disassociated].
2016-03-11 05:28:20,345 INFO  org.apache.flink.yarn.YarnFlinkResourceManager                - Container ResourceID{resourceId='container_1440768826963_0010_01_000002'} failed. Exit status: 143
2016-03-11 05:28:20,346 INFO  org.apache.flink.yarn.YarnFlinkResourceManager                - Diagnostics for container ResourceID{resourceId='container_1440768826963_0010_01_000002'} in state COMPLETE : exitStatus=143 diagnostics=Container killed on request. Exit code is 143
Container exited with a non-zero exit code 143
Killed by external signal

2016-03-11 05:28:20,346 INFO  org.apache.flink.yarn.YarnFlinkResourceManager                - Total number of failed containers so far: 1
2016-03-11 05:28:20,346 INFO  org.apache.flink.yarn.YarnFlinkResourceManager                - Requesting new TaskManager container with 1024 megabytes memory. Pending requests: 1
2016-03-11 05:28:20,346 INFO  org.apache.flink.yarn.YarnJobManager                          - Task manager akka.tcp://flink@127.0.0.1:48995/user/taskmanager terminated.
2016-03-11 05:28:20,347 INFO  org.apache.flink.runtime.instance.InstanceManager             - Unregistered task manager akka.tcp://flink@127.0.0.1:48995/user/taskmanager. Number of registered task managers 0. Number of available slots 0.
2016-03-11 05:28:21,859 WARN  Remoting                                                      - Tried to associate with unreachable remote address [akka.tcp://flink@127.0.0.1:48995]. Address is now gated for 5000 ms, all messages to this address will be delivered to dead letters. Reason: Connection refused: /127.0.0.1:48995
2016-03-11 05:28:25,850 INFO  org.apache.flink.yarn.YarnFlinkResourceManager                - Received new container: container_1440768826963_0010_01_000003 - Remaining pending container requests: 0
2016-03-11 05:28:25,850 INFO  org.apache.flink.yarn.YarnFlinkResourceManager                - Launching TaskManager in container ResourceID{resourceId='container_1440768826963_0010_01_000003'} on host quickstart.cloudera
2016-03-11 05:28:25,859 INFO  org.apache.flink.yarn.YarnFlinkResourceManager                - Received new container: container_1440768826963_0010_01_000004 - Remaining pending container requests: 0
2016-03-11 05:28:25,859 INFO  org.apache.flink.yarn.YarnFlinkResourceManager                - Returning excess container container_1440768826963_0010_01_000004
2016-03-11 05:28:26,354 INFO  org.apache.flink.yarn.YarnFlinkResourceManager                - Container ResourceID{resourceId='container_1440768826963_0010_01_000004'} completed successfully with diagnostics: Container released by application
2016-03-11 05:28:28,034 INFO  org.apache.flink.runtime.instance.InstanceManager             - Registered TaskManager at quickstart (akka.tcp://flink@127.0.0.1:36513/user/taskmanager) as 22bdcc2288142acf37c723fa2930d447. Current number of registered hosts is 1. Current number of alive task slots is 1.
```

rmetzger commented:
Overall, the change looks good. I'm happy to see that the YARN tests haven't been touched much. As long as these tests are present, most of the features users need should work properly.
Also my tests on a secured virtual machine were successful.

I don't have the time to check in detail if all the message sequences and different procedures are implemented correctly but given that everything worked when I tested it, I'm positive that we can merge the change soon.

I think a quick review by @StephanEwen and @tillrohrmann would be good.

mxm commented:
Thank you for checking out the PR @rmetzger! Great to see it working in a cluster environment with Task Manager failures. I didn't add anything to the documentation (apart from JavaDocs) because this is an internal feature. It shouldn't affect the user apart form config parameters. I agree that these have to be documented exhaustively.

rmetzger commented:
I didn't test the change on a physical, real distributed cluster. Only on a virtual machine from cloudera.
We should test the code on a real setup as well before merging.

StephanEwen commented:
I have started looking into this, but will need a bit more time...

Since there are already good sketches (in the attached PDF), it would be great to add these sketches to the "internals" docs, with a few lines of explanation. It might actually be interesting to a bunch of users...

tillrohrmann commented:
I would also need some more time for review if my review is requested.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/ContaineredTaskManagerParameters.java:
```diff
@@ -0,0 +1,182 @@
+
+	
+	public ContaineredTaskManagerParameters(
+		long totalContainerMemoryMB, long taskManagerHeapSizeMB,
```
What's the pattern of line breaks you're following when wrapping parameter lists?

mxm replied:
Line break, followed by tab, then for the first line fill with parameters until 100 chars, where one tab equals 4 chars. For the following lines, only break line, insert tab, then one parameter. Not sure if it is consistent though. Sometimes I simply break lines followed by tab and one parameter. What's the pattern you're using?

tillrohrmann replied:
I think I would place every parameter in one line or on every line multiple parameters. I don't think that multiple parameters on the first line followed by a single parameter on each following line is the most intuitive pattern.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/BootstrapTools.java:
```diff
@@ -0,0 +1,343 @@
+				String jobManagerHostname,
+				int jobManagerPort,
+				int numSlots,
+				FiniteDuration registrationTimeout) {
```
What is your standard indentation level for line wrapped parameter lists?

mxm replied:
One tab, but it seems sometimes I prefer two or more.

tillrohrmann replied:
I was just wondering because I noted that you did it differently in different source files.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/ContaineredTaskManagerParameters.java:
```diff
@@ -0,0 +1,182 @@
+			ConfigConstants.CONTAINERED_HEAP_CUTOFF_MIN,
+			ConfigConstants.DEFAULT_YARN_HEAP_CUTOFF);
+
+		if (memoryCutoffRatio > 1 || memoryCutoffRatio < 0) {
```
Do we allow a `memoryCutoffRatio` of `1`?

mxm replied:
I suppose it doesn't matter, even 0.9999 will probably fail hard. However, it is not consistent with the error message that states that the ratio should be between 0 and 1.

mxm replied:
Fixed limits.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/FlinkResourceManager.java:
```diff
@@ -0,0 +1,801 @@
+		}
+		// framework specific initialization
+		try {
+			initialize();
```
It probably does not make much sense to initialize the resource manager if an error occurred while starting the leader retrieval service.

mxm replied:
Good catch. I consolidated this into one try/catch block.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/FlinkResourceManager.java:
```diff
@@ -0,0 +1,801 @@
+
+			@Override
+			public void onComplete(Throwable failure, Object msg) {
+				if (msg != null) {
```
I think it would be better to first check whether `failure != null` which indicates a failure case. Furthermore, I would log this `Throwable` with the failure message.

mxm replied:
The docs state: 
> In the case of success then "failure" will be null, and in the case of failure the "success" [msg] will be null.

It seems like in this case it doesn't matter if we check `msg` or `failure`. However, I've added the `failure` to the logging.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/FlinkResourceManager.java:
```diff
@@ -0,0 +1,796 @@
+				List<ResourceID> toHandle = new ArrayList<>(workers.size());
+				for (ResourceID resourceID : workers) {
+					toHandle.add(resourceID);
+				}
```
`toHandle.addAll(workers)` should have a better performance.

mxm replied:
After a brief look into `addAll(..)`, I'm not sure. It seems that a simple add is more efficient if the ArrayList is pre-initialized to the right size. `addAll(..)` performs an array copy first. In terms of simplicity of code I agree that addAll looks simpler. Changed.

tillrohrmann replied:
No it is not right that the for loop is more efficient, because array copy won't create or change the size of the existing array if it is large enough to store the source array.

mxm replied:
I didn't see where the `add` method grows the ArrayList in this case where the list is already correctly sized.

mxm replied:
In this particular case it is more efficient but it doesn't really matter.

tillrohrmann replied:
I never said it did. In fact both variants won't grow the underlying array. However, copying arrays via `System.arraycopy` is faster than an element by element copying for loop.

mxm replied:
Okay I get your point. Still, it is not a performance critical at all here. And even looping and inserting is very fast.

tillrohrmann replied:
Sure, but why not choosing the better alternative if it's less code and faster (even if it's not critical but we don't have to waste cycles if it comes for free)

mxm replied:
Yes sure. I already changed it after your initial comment. I just wanted to point out that it has a very minor performance impact (if at all). In terms of clarity of the code, I totally agree that `addAlll` looks much nicer.

tillrohrmann replied:
Just FYI, I was curious to see the effects of arraycopy vs. for loop and tested it with copying an 8 element int array. The for loop variant was consistently 50% slower than the arraycopy variant (on average).

mxm replied:
Wow, wouldn't have thought that the pointer arithmetic eats up that many CPU cycles. Valuable lesson learned here.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/FlinkResourceManager.java:
```diff
@@ -0,0 +1,796 @@
+					// put the consolidated TaskManagers into our bookkeeping
+					for (WorkerType worker : consolidated) {
+						registeredWorkers.put(worker, worker);
+						toHandle.remove(worker);
```
I think a `Set` would be more efficient here since removing elements from an `ArrayList` has complexity `O(n)`.

mxm replied:
Agreed and changed but I don't think that we are at a performance critical execution path here.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/FlinkResourceManager.java:
```diff
@@ -0,0 +1,796 @@
+		ActorSystem actorSystem,
+		LeaderRetrievalService leaderRetriever,
+		Class<? extends FlinkResourceManager<?>> resourceManagerClass
+	) {
```
one line break too much

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/FlinkResourceManager.java:
```diff
@@ -0,0 +1,796 @@
+		LeaderRetrievalService leaderRetriever,
+		Class<? extends FlinkResourceManager<?>> resourceManagerClass,
+		String resourceManagerActorName
+	) {
```
Same here

mxm replied:
I think it looks cleaner that way with the parameters.

tillrohrmann replied:
But it is neither consistent with the rest of the code base nor with other parts of this PR.

mxm replied:
As far as I know there is no consistent style in the code base. I value readability more than a consistent style (and it is really just about a new line here). I'm happy to adjust it to a particular code style according to the style guide but it currently doesn't state anything about constructors.

tillrohrmann replied:
You're right that we don't have a defined coding style. However, we always try to motivate people to stick a bit to the existing code in the code base.

I mean, personally, I don't have a strong preference. However, I'm wondering why you don't do it like this everywhere if it gives a better readability.

mxm replied:
I just don't feel so strong about it but I can see how it can drive other people nuts. I'll do another pass to fix some of these style inconsistencies.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/messages/RegisterInfoMessageListenerSuccessful.java:
```diff
@@ -0,0 +1,69 @@
+
+	@Override
+	public int hashCode() {
+		return 2018741654;
```
This class and the `RegisterInfoMessageListener` message have the same hash value.

mxm replied:
This is fine. Two objects may have the same hash value. That doesn't imply they're equal. Still, changed it for clarity.

tillrohrmann replied:
Sure, but it would cause more hash collisions which is not good.

Btw: I was wondering why you implemented `equals` and `hashCode` only for a subset of messages but not for all? Does not appear to be consistent.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/messages/RegisterResourceManagerSuccessful.java:
```diff
@@ -0,0 +1,79 @@
+	/** The list of registered TaskManagers that the JobManager currently knows */
+	private final Collection<ResourceID> currentlyRegisteredTaskManagers;
+
+
```
Two line breaks

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/messages/SetWorkerPoolSize.java:
```diff
@@ -0,0 +1,63 @@
+
+	public SetWorkerPoolSize(int numberOfWorkers) {
+		if (numberOfWorkers < 0) {
+			throw new IllegalArgumentException();
```
Exception message could be helpful

mxm replied:
Added

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/messages/SetWorkerPoolSize.java:
```diff
@@ -0,0 +1,63 @@
+	
+	@Override
+	public int hashCode() {
+		return 31 + numberOfWorkers;
```
Why do you add `31` here?

mxm replied:
Kind of arbitrary I agree. `numberOfWorkers` is enough.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/messages/ShutdownClusterAfterJob.java:
```diff
@@ -0,0 +1,46 @@
+	
+	private final JobID jobId;
+
+	
```
two linebreaks

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/messages/ShutdownClusterAfterJob.java:
```diff
@@ -0,0 +1,46 @@
+ * Message sent to the cluster framework master to signal it that the cluster
+ * should be shut down upon completion of a certain job.
+ */
+public class ShutdownClusterAfterJob implements java.io.Serializable {
```
This class is nowhere used in the code. What's its purpose?

mxm replied:
This should replace the Yarn shutdown logic. I've integrated it now to replace `StopAMAfterJob`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/messages/GetClusterStatus.java:
```diff
@@ -0,0 +1,68 @@
+ * This message signals the resource master to check how many TaskManagers are 
+ * desired, how many are available, and to trigger adjustments if needed.
+ */
+public class GetClusterStatus implements java.io.Serializable {
```
Class is not used according to IntelliJ.

mxm replied:
Same here. General mechanism. Implemented now to replace `PollYarnClusterStatus`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/messages/GetClusterStatusResponse.java:
```diff
@@ -0,0 +1,54 @@
+
+import java.io.Serializable;
+
+public class GetClusterStatusResponse implements Serializable {
```
Class is not used

mxm replied:
Thanks. Also integrated and replaces `FlinkYarnClusterStatus`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/messages/StopClusterSuccessful.java:
```diff
@@ -0,0 +1,48 @@
+
+	public static StopClusterSuccessful get() {
+		if (INSTANCE == null) {
+			INSTANCE = new StopClusterSuccessful();
```
Concurrency issue here: Multiple threads might create more than one singleton here.

mxm replied:
Good catch.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/standalone/HeartbeatAcknowledgement.java:
```diff
@@ -0,0 +1,69 @@
+import org.apache.flink.runtime.messages.RequiresLeaderSessionID;
+
+/**
+ * This message is teh successful response to a heartbeat. 
```
typo: the

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/standalone/StandaloneResourceManager.java:
```diff
@@ -0,0 +1,123 @@
+	@Override
+	protected void fatalError(String message, Throwable error) {
+		log.error("FATAL ERROR IN RESOURCE MANAGER: " + message, error);
+		LOG.error("Shutting down process");
```
Why do we use two different logger variables here?

mxm replied:
Yep, will change that. There is the Actor logger and the RM logger. No need to have both.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceID.java:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Class for Resource Ids assigned at the FlinkResourceManager.
+ */
+public class ResourceID implements Serializable {
```
Why don't we use `AbstractID` here?

mxm replied:
In standalone mode, this is just a random String but in Yarn mode, this is the container id assigned at the framework. Thus, not random.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/messages/MessageDecorator.java:
```diff
@@ -21,7 +21,7 @@
  * Interface for message decorators
  */
-public interface MessageDecorator {
+public interface MessageDecorator extends java.io.Serializable {
```
Good catch :-)

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -886,10 +957,42 @@ class JobManager(
         log.info(s"Task manager ${taskManager.path} wants to disconnect, because $msg.")
 
-        instanceManager.unregisterTaskManager(taskManager, false)
+          instanceManager.unregisterTaskManager(taskManager, false)
```
Formatting regression

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -886,10 +957,42 @@ class JobManager(
+          // inform rm
+          rm ! decorateMessage(msg)
+
+          respondTo ! decorateMessage(StopClusterSuccessful.get())
```
Maybe rename `get` to `getInstance`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/standalone/HeartbeatAcknowledgement.java:
```diff
@@ -0,0 +1,69 @@
+/**
+ * This message is teh successful response to a heartbeat. 
+ */
+public class HeartbeatAcknowledgement implements RequiresLeaderSessionID, java.io.Serializable {
```
Class is not used

mxm replied:
You're looking at an old commit. I already removed this class but didn't want to force-push.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/standalone/HeartbeatMessage.java:
```diff
@@ -0,0 +1,60 @@
+/**
+ * Heartbeat message send by a standalone TaskManager to the resource master.
+ */
+public class HeartbeatMessage implements RequiresLeaderSessionID, Serializable {
```
Class is not used

mxm replied:
You're looking at an old commit. I already removed this class but didn't want to force-push.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/standalone/StandaloneMode.java:
```diff
@@ -16,24 +16,21 @@
+/**
+ * The startup mode for the standalone setup,
+ */
+public enum StandaloneMode {
```
Class is not used

mxm replied:
You're looking at an old commit. I already removed this class but didn't want to force-push.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -312,59 +323,125 @@ class JobManager(
+          // we should ditch the current resource manager
+          log.debug(s"Disconnecting resource manager $rm.")
+          // send the old one a disconnect message
+          rm ! decorateMessage(new TriggerRegistrationAtJobManager(self))
```
What happens if this message never reaches the ResourceManager? Is there a mean that the RM can detect that it lost connection to the JM?

mxm replied:
You're right that the message can be lost. Apart from the leader election service there is currently no mechanism to detect a lost connection from RM to the JM.

The above code is part of a special case where the RM doesn't reply to a TM registration and the JM decides to disconnect the RM. I suppose we should keep retrying to send the TriggerRegistrationAtJobManager until we receive a registration of a RM.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -312,59 +323,125 @@ class JobManager(
+
+      taskManager ! decorateMessage(
+        RefuseRegistration(
+          ExceptionUtils.stringifyException(new IllegalStateException(
```
Can we change `RefuseRegistration` such that we can simply give a `Throwable` as parameter instead of using this stringifyException tool?

mxm replied:
Sure. You want to stringify it in the `RefuseRegistration` constructor, right?

mxm replied:
Then we can't use a case class anymore. I think I'd rather keep it like it is. I think it was on purpose not to pass Throwables because there might be different classloaders at the jobmanager and the task manager. I didn't change the message. It is just like that in the master.

tillrohrmann replied:
What do you mean with different class loaders? I think it should be perfectly fine to have a
```
case class RefuseRegistration(t: Throwable)
```
because all `Throwables` inserted here should be known to the system class loader.

mxm replied:
All I'm saying is that I'm not perfectly aware of subtle bugs this change would introduce. It seems like String was deliberately chosen to avoid ClassNotFoundExceptions. (Think of different Flink versions talking to each other) Thus, I'd rather not change it.

tillrohrmann replied:
I think it's an legacy artifact. Moreover, differing versions of an `IllegalStateException` will be your least problem when you have different version of Flink talking to each other.

mxm replied:
Ok, changing to send the `Throwable` directly.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -886,10 +957,42 @@ class JobManager(
+          // retry
+          context.system.scheduler.scheduleOnce(
+            2 seconds, self, decorateMessage(msg)
+          )(context.dispatcher)
```
What if we will never establish a connection to the RM again for some reason? Wouldn't that mean that we will never shutdown the JM?

tillrohrmann replied:
What if there was never a RM started? Would that make a difference?

mxm replied:
You're right. There should be an upper bound for the number of retries. If there was never a RM, then is behaves the same.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/messages/Messages.scala:
```diff
@@ -29,22 +29,22 @@ object Messages {
+    *
+    * @param reason The reason for disconnecting, to be displayed in log and error messages.
+    */
+  case class Disconnect(reason: String) extends RequiresLeaderSessionID
```
I don't see the reason for these changes here.

mxm replied:
You're right, there is no need to change the order here. I'll revert the changes.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/messages/RegistrationMessages.scala:
```diff
@@ -36,7 +38,7 @@ object RegistrationMessages {
    * Triggers the TaskManager to attempt a registration at the JobManager.
    *
-   * @param jobManagerURL Akka URL to the JobManager
+   * @param jobManagerURL Akka URL to the JobManager to ask for the JobManager
```
I don't understand the comment here. Admittedly, before it was not much better.

mxm replied:
For me it is the other way around. Maybe change this to 
>Akka URL to the JobManager to ask for the JobManager actor

?

mxm replied:
Reverted the comment.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/messages/RegistrationMessages.scala:
```diff
@@ -87,7 +91,7 @@ object RegistrationMessages {
 
   /**
-   * Denotes the unsuccessful registration of a task manager at the job manager. This is the
+   * Denotes the unsuccessful registration of a task manager at the JobManager. This is the
```
This change and the change in line 74 don't seem to be consistent.

mxm replied:
I've made it consistent.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/minicluster/LocalFlinkMiniCluster.scala:
```diff
@@ -211,6 +238,13 @@ class LocalFlinkMiniCluster(
       JobManager.JOB_MANAGER_NAME
     }
   }
+  protected def getResourceManagerName(index: Int): String = {
```
line break missing before this line

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/minicluster/LocalFlinkMiniCluster.scala:
```diff
@@ -211,6 +238,13 @@ class LocalFlinkMiniCluster(
+    } else {
+      FlinkResourceManager.RESOURCE_MANAGER_NAME
+    }
+  }
```
After this line there is also a line break missing

mxm replied:
Corrected, but not part of this diff.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -592,19 +605,18 @@ class TaskManager(
             case t: Throwable =>
               killTaskManagerFatal(
-                "Unable to start TaskManager components after registering at JobManager", t)
+                "Unable to start TaskManager components and associate with the JobManager ", t)
```
Why the whitespace at the end of the error message?

mxm replied:
Not necessary. Removing whitespace.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -592,19 +605,18 @@ class TaskManager(
-      // we are already registered at that specific JobManager - duplicate answer, rare cases
+      // we are already registered at this ResourceManager - duplicate answer, rare cases
       case AlreadyRegistered(id, blobPort) =>
         val jobManager = sender()
```
Is this the JM or the RM address which we store here in `jobManager`?

mxm replied:
This is an artifact of refactoring the original approach that I followed where the TaskManager would register at the ResourceManager. Reverting.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -613,12 +625,11 @@ class TaskManager(
           } else {
             log.warn(s"Received 'AlreadyRegistered' message from " +
-              s"JobManager ${jobManager.path}, even through TaskManager is currently " +
+              s"ResourceManager ${jobManager.path}, even through TaskManager is currently " +
```
Does `jobManager` contain the `ActorRef` of the RM or the JM?

mxm replied:
This is an artifact of refactoring the original approach that I followed where the TaskManager would register at the ResourceManager. Reverting.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -632,34 +643,32 @@ class TaskManager(
       case RefuseRegistration(reason) =>
         if (currentJobManager.isEmpty) {
-          log.error(s"The registration at JobManager $jobManagerAkkaURL was refused, " +
+          log.error(s"The registration at ResourceManager $jobManagerAkkaURL was refused, " +
```
RM vs. JM conflict?

mxm replied:
This is an artifact of refactoring the original approach that I followed where the TaskManager would register at the ResourceManager. Reverting.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -928,11 +934,6 @@ class TaskManager(
-    // de-register from the JobManager (faster detection of disconnect)
-    currentJobManager foreach {
-      _ ! decorateMessage(Disconnect(s"TaskManager ${self.path} is disassociating"))
-    }
```
Why don't we need this anymore?

mxm replied:
Also an artifact of the original approach. Reverting.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -1823,8 +1836,8 @@ object TaskManager {
-   * @throws java.io.IOException Thrown, if the lookup fails.
-   * @return The ActorRef to the TaskManager
+    * @throws java.io.IOException Thrown, if the lookup fails.
+    * @return The ActorRef to the TaskManager
```
formatting regression

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -2066,8 +2079,7 @@ object TaskManager {
    * @param errorMessage The optional custom error message to append to the exception message.
-   *
-   * @throws IllegalConfigurationException Thrown if the condition is violated.
+    * @throws IllegalConfigurationException Thrown if the condition is violated.
```
Wrong indentation

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java:
```diff
@@ -0,0 +1,109 @@
+
+		String desig1 = deprecatedPrefix1 + "var";
+		String desig2 = deprecatedPrefix2 + "env";
+		String desig3 = deprecatedPrefix2 + "x";
```
Is it intended that you used the deprecatedPrefix for the designated keys?

mxm replied:
Nope. Good catch.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/ResourceManagerITCase.java:
```diff
@@ -0,0 +1,199 @@
+import static org.junit.Assert.assertTrue;
+
+
+public class ResourceManagerITCase {
```
Extending the `TestLogger` will give use nice outputs which test case is currently being run in the logs.

mxm replied:
Thanks. Adding.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/ResourceManagerITCase.java:
```diff
@@ -0,0 +1,199 @@
+	public void testResourceManagerReconciliation() {
+
+		new JavaTestKit(system){{
+		new Within(duration("10 seconds")) {
```
indentation

mxm replied:
This is intentionally. Readability sometimes goes over a strict style for me.

tillrohrmann replied:
Then I guess we have a different understanding of readability.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/ResourceManagerITCase.java:
```diff
@@ -0,0 +1,199 @@
+						} catch (InterruptedException e) {
+							// try again
+						}
+					}
```
Can't the shutdown of the `ActorSystem` be problematic if JUnit does not execute the tests in the order they are defined in this file? Assume that you first run `testClusterShutdown` and then the rest of the test cases.

mxm replied:
Actually not because the test cases uses `@Before` and `@After` to bring up an ActorSystem for each test.

tillrohrmann replied:
True, I mistook that for a `BeforeClass` and `AfterClass`. Wouldn't it then make sense to put all tests which can work using the same `ActorSystem` in one class using `AfterClass`/`BeforeClass` and leave the tests which require a fresh `ActorSystem` here? Bringing up an `ActorSystem` is quite costly. Especially wrt to long test runtimes this could be helpful.

mxm replied:
Done. Moved that test to a new class and prefixed the actor names in this class such that we can start several job manager actors.

tillrohrmann commented on flink-scala/src/main/scala/org/apache/flink/api/scala/ExecutionEnvironment.scala:
```diff
@@ -40,6 +40,7 @@ import org.apache.hadoop.mapred.{FileInputFormat => MapredFileInputFormat, Input
 import org.apache.hadoop.mapreduce.lib.input.{FileInputFormat => MapreduceFileInputFormat}
 import org.apache.hadoop.mapreduce.{InputFormat => MapreduceInputFormat, Job}
 
+import scala.annotation.implicitNotFound
```
Unused import

mxm replied:
Ah true. Why is it actually that our Scala Checkstyle does not detect unused imports? Not supported or not configured?

tillrohrmann replied:
Good question, it's hopefully the latter case.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -60,8 +63,8 @@
-		float memoryCutoffRatio = conf.getFloat(ConfigConstants.YARN_HEAP_CUTOFF_RATIO, ConfigConstants.DEFAULT_YARN_HEAP_CUTOFF_RATIO);
-		int minCutoff = conf.getInteger(ConfigConstants.YARN_HEAP_CUTOFF_MIN, ConfigConstants.DEFAULT_YARN_MIN_HEAP_CUTOFF);
+		float memoryCutoffRatio = conf.getFloat(ConfigConstants.YARN_HEAP_CUTOFF_RATIO, 0.25f);
+		int minCutoff = conf.getInteger(ConfigConstants.YARN_HEAP_CUTOFF_MIN, 600);
```
Why not using the default values defined in the `ConfigConstants`?

mxm replied:
Originally, those were going to be replaced by generalized keys and values. However, we need default values for every resource management system. So I'll use the generalized keys with the default yarn values here. Like so:

```java
		float memoryCutoffRatio = conf.getFloat(ConfigConstants.CONTAINERED_HEAP_CUTOFF_RATIO,
			ConfigConstants.DEFAULT_YARN_HEAP_CUTOFF_RATIO);
		int minCutoff = conf.getInteger(ConfigConstants.CONTAINERED_HEAP_CUTOFF_MIN,
			ConfigConstants.DEFAULT_YARN_HEAP_CUTOFF);
```

The deprecated keys are replaced beforehand using

```java
        BootstrapTools.substituteDeprecatedConfigKey(conf,
            ConfigConstants.YARN_HEAP_CUTOFF_RATIO, ConfigConstants.CONTAINERED_HEAP_CUTOFF_RATIO);
        BootstrapTools.substituteDeprecatedConfigKey(conf,
            ConfigConstants.YARN_HEAP_CUTOFF_MIN, ConfigConstants.CONTAINERED_HEAP_CUTOFF_MIN);
```

tillrohrmann replied:
I was actually referring to why not using `ConfigConstants.DEFAULT_YARN_HEAP_CUTOFF` instead of `600`?

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/ResourceManagerITCase.java:
```diff
@@ -0,0 +1,199 @@
+	@Test
+	public void testClusterShutdown() {
+		new JavaTestKit(system){{
+			new Within(duration("30 seconds")) {
```
Here `Within` is indented.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationMasterRunner.java:
```diff
@@ -0,0 +1,601 @@
+		if (webMonitor != null) {
+			try {
+				webMonitor.stop();
+			} catch (Throwable ignored) {}
```
Would be maybe good to log the error.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationMasterRunner.java:
```diff
@@ -0,0 +1,601 @@
+			if (webMonitor != null) {
+				try {
+					webMonitor.stop();
+				} catch (Throwable ignored) {}
```
Would be maybe good to log the error.

mxm replied:
Doesn't that tend to hide the original error we caught? The regular shutdown should of course log errors of `stop()`.

tillrohrmann replied:
It could be then logged as a warning maybe.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -23,7 +23,7 @@ import java.net.{BindException, ServerSocket, UnknownHostException, InetAddress,
 import java.util.concurrent.{TimeUnit, ExecutorService}
 
-import akka.actor.Status.Failure
+import akka.actor.Status.{Success, Failure}
```
`Success` unused

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -39,24 +39,32 @@ import org.apache.flink.runtime.blob.BlobServer
 import org.apache.flink.runtime.executiongraph.restart.{RestartStrategy, RestartStrategyFactory}
 import org.apache.flink.runtime.executiongraph.{ExecutionGraph, ExecutionJobVertex}
-import org.apache.flink.runtime.instance.{AkkaActorGateway, InstanceManager}
+import org.apache.flink.runtime.instance.{HardwareDescription, InstanceConnectionInfo,
```
`HardwareDescription` and `InstnaceConnectionInfo` unused

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkResourceManager.java:
```diff
@@ -0,0 +1,647 @@
+	private List<Container> getContainersFromPreviousAttempts(RegisterApplicationMasterResponse response) {
+		try {
+			Method m = RegisterApplicationMasterResponse.class
+				.getMethod("getContainersFromPreviousAttempts");
```
Why did you remove the `RegisterApplicationMasterResponseReflector`? With the current implementation you will use reflection every time you call `getContainersFromPreviousAttempts`. This is inefficient and not necessary since `RegisterApplicationMasterResponse.class` won't change.

mxm replied:
It changes because everything is Java now. Before, the RegisterApplicationMasterResponseReflector was in YarnJobManager.scala.

I tend to think this is premature optimization, but I'll add it again in Java because we already had it before.

tillrohrmann commented on flink-yarn/src/main/scala/org/apache/flink/yarn/ApplicationClient.scala:
```diff
@@ -115,7 +115,7 @@ class ApplicationClient(
 
           jobManager ! decorateMessage(
-            RegisterApplicationClient
+            RegisterInfoMessageListener.get()
```
`getInstance`?

tillrohrmann commented on flink-yarn/src/main/scala/org/apache/flink/yarn/ApplicationClient.scala:
```diff
@@ -148,7 +148,7 @@ class ApplicationClient(
           WAIT_FOR_YARN_INTERVAL,
           yarnJobManager.get,
-          decorateMessage(PollYarnClusterStatus))
+          decorateMessage(GetClusterStatus.get()))
```
`getInstance`?

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -312,59 +323,121 @@ class JobManager(
+            case scala.util.Success(response) =>
+              // the resource manager is available and answered
+              self ! response
+            case scala.util.Failure(t) =>
```
Could be good to log the failure.

mxm replied:
Done (doesn't change the diff).

tillrohrmann commented:
Great work @mxm! I really like the new architecture :-)

I had some minor comments. The only thing which is important to fix is that the `JobManager` can terminate if it is not connected to a `FlinkResourceManager`. Apart from that, I think we should merge it soon so that it can get some exposure.

mxm commented:
Thanks a lot @tillrohrmann for taking the time to look into the bulk of code. I have already eagerly addressed most of your comments in the additional commits I pushed. Next, I'll revise the cluster shutdown logic and the re-connect in case of unresponsiveness of the resource manager.

mxm commented:
I've incorporated the changes and the tests pass. I would like to merge the pull request. Please let me know if there are still pending code reviews.

StephanEwen commented:
Continuing my review now, but don't block this on me. If I find anything crucial, I will open a pull request against master if it is merged by then..
